### PR TITLE
apply OpenBLAS_jll v0.3.23+4 patch

### LINKED ISSUE #1051: Non-deterministic `mul!` of `ComplexF32` arrays under Apple Silicon
`mul!` of `ComplexF32` arrays is subtly broken under Apple Silicon in Julia 1.9 and 1.10

Here is a MWE
```
using LinearAlgebra

function test()
    C = ComplexF32
    mat = zeros(C, 1, 1)
    for _ in 1:100
        v = [C(1-0.2im) C(2+0.3im)]
        mul!(mat, v, v', C(1+im), 1)
    end
    return mat
end

allequal(test() for _ in 1:10000)
```

This should give always `true`, but it randomly gives `false` in the version/OSs above. The deviations of "bad" runs of `test()` (a small percentage of the 10000) is small, but well within the precision of `ComplexF32`. 

This was tested on a Macbook Pro M1 (fails) and a Xeon server under Linux (does not fail). Can someone reproduce?

Notably, current master does not have this problem on macos. I'll try to bisect to see when it was fixed, in case we can backport the patch.

### LINKED ISSUE #53054: Non-deterministic `mul!` of `ComplexF32` arrays under Apple Silicon
`mul!` of `ComplexF32` arrays is subtly broken under Apple Silicon in Julia 1.9 and 1.10

Here is a MWE
```
using LinearAlgebra

function test()
    C = ComplexF32
    mat = zeros(C, 1, 1)
    for _ in 1:100
        v = [C(1-0.2im) C(2+0.3im)]
        mul!(mat, v, v', C(1+im), 1)
    end
    return mat
end

allequal(test() for _ in 1:10000)
```

This should give always `true`, but it randomly gives `false` in the version/OSs above. The deviations of "bad" runs of `test()` (a small percentage of the 10000) is small, but well within the precision of `ComplexF32`. 

This was tested on a Macbook Pro M1 (fails) and a Xeon server under Linux (does not fail). Can someone reproduce?

Notably, current master does not have this problem on macos. I'll try to bisect to see when it was fixed, in case we can backport the patch.

### PR DESCRIPTION
**Pull request by pablosanjose**
Closes JuliaLang/LinearAlgebra.jl#1051

CC: @giordano @ViralBShah

### CODE REVIEW

File: stdlib/OpenBLAS_jll/Project.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 name = "OpenBLAS_jll"
 uuid = "4536629a-c528-5b80-bd46-f80d51c5b363"
-version = "0.3.23+2"
+version = "0.3.23+3"
```
Comment (giordano) [inline stdlib/OpenBLAS_jll/Project.toml:3]: I believe you want the +4 build, not the +3 one: https://github.com/JuliaRegistries/General/pull/99646

Comment (pablosanjose) [reply stdlib/OpenBLAS_jll/Project.toml]: Oh, right! I was wondering what established the patch number...

**Comment by pablosanjose** [comment]
Two questions:
(1) is CI working correctly? If so, I cannot make heads or tails of the test failures
(2) should we add the MWE of https://github.com/JuliaLang/julia/issues/53054 as a test?

**Comment by giordano** [comment]
First of all, have you verified the bug is indeed gone on this branch? Regarding a regression test, that'd be a good idea, a test only on this branch would be quite useless as it'd be lost forever in future releases. You want to submit a separate PR for `master`, perhaps to be backported to 1.10 (but the backport doesn't seem massively useful).

**Review by ViralBShah — APPROVED** [review]

**Review by ViralBShah — CHANGES_REQUESTED** [review]
On 1.10 branch, we also need to add the patch for a source based build. So the patch should also get added to `deps/patches`, and it should be applied in `deps/openblas.mk`.

**Comment by ViralBShah** [comment]
The best way to confirm the bug is indeed fixed is to use the build artifact produced by this PR:

https://buildkite.com/organizations/julialang/pipelines/julia-release-1-dot-10/builds/191/jobs/018d4778-8e0d-4db8-8f73-bd65ca4992ae/artifacts/018d477e-8eb1-4b6f-a13c-df29e0d4c1fe

**Comment by imciner2** [comment]
> (2) should we add the MWE of https://github.com/JuliaLang/julia/issues/53054 as a test?

I'd say the regression test should at a minimum go into master (we added one for the other OpenBLAS patch we are using for M1 just recently in https://github.com/JuliaLang/julia/pull/53059). That should be a separate PR though.

**Comment by pablosanjose** [comment]
> First of all, have you verified the bug is indeed gone on this branch?

I can confirm that for me this branch builds from source using BinaryBuilder, and the result is free of the bug.

**Comment by pablosanjose** [comment]
Regarding the test failures, at least some of them are also appearing in buildkite runs of other branches, such as [avi/53062](https://buildkite.com/julialang/julia-release-1-dot-10/builds?branch=avi%2F53062), so they should be unrelated to the patch.

> Regarding a regression test, that'd be a good idea, but a test only on this branch would be quite useless as it'd be lost forever in future releases. You want to submit a separate PR for master, perhaps to be backported to 1.10 (but the backport doesn't seem massively useful).

> I'd say the regression test should at a minimum go into master (we added one for the other OpenBLAS patch we are using for M1 just recently in https://github.com/JuliaLang/julia/pull/53059). That should be a separate PR though.

I'll add a test here and submit it also to master as a separate PR.

**Comment by pablosanjose** [comment]
There is a potential issue I'm trying to understand: this patch is for https://github.com/JuliaRegistries/General/pull/99646, which is indeed v0.3.23+4, but OpenBLAS32 version (https://github.com/JuliaRegistries/General/pull/99647) is v0.3.23+3. Is this a problem?

**Comment by giordano** [comment]
> Is this a problem?

No.

File: stdlib/OpenBLAS_jll/test/runtests.jl [hunk 2]
```diff
@@ -14,4 +14,17 @@ end
 
 @testset "OpenBLAS_jll" begin
     @test dlsym(OpenBLAS_jll.libopenblas_handle, @blasfunc(openblas_set_num_threads); throw_error=false) != nothing
+    @testset "deterministic mul!" begin
```
Comment (giordano) [inline stdlib/OpenBLAS_jll/test/runtests.jl:17]: I don't think this test should be here, this is meant to test only very basic functionalities of the jll, like opening the library and such.

Comment (pablosanjose) [reply stdlib/OpenBLAS_jll/test/runtests.jl:17]: Oh, this was [a suggestion](https://github.com/JuliaLang/julia/pull/53079#issuecomment-1913250911) from @KristofferC

Comment (pablosanjose) [reply stdlib/OpenBLAS_jll/test/runtests.jl]: reverted

**Comment by KristofferC** [comment]
So this should be merged in addition to https://github.com/JuliaLang/julia/pull/52755/commits/3c738736ebee9e2aa25f1203aee3ae0c85e5e80a?

**Comment by KristofferC** [comment]
Test error is for another change on the backport branch.

**Pull request merged**